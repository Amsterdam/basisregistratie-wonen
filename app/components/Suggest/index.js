import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Link from 'components/Link';

const Strong = styled.strong`
  padding-left: 4px;
`;

const StyledLink = styled(Link)`
  margin-left: 4px !important;
`;

const Ul = styled.ul`
  background: white;
  border-top: none;
  border: #767676 solid 1px;
  width: 100%;

  li:first-of-type {
    margin-top: 8px !important;
  }

  @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    a {
      height: 26px;
    }
  }
`;

/* eslint-disable indent,react/jsx-indent */
const Suggest = forwardRef(({ className, items, onSelect }, ref) => {
  if (!items) return null;

  const numItems = Object.keys(items).length;

  if (!numItems) return null;

  return (
    <Ul className={`${className} links`} ref={ref} numItems={numItems}>
      {Object.keys(items).map(key => (
        <li key={key}>
          <Strong>{key}</Strong>

          <ul className="links">
            {items[key].map(item => (
              <li key={`${item.name}_${item.vboId || item.ligId || item.brkId}`}>
                <StyledLink
                  href="/"
                  data-vbo-id={item.vboId}
                  data-lig-id={item.ligId}
                  data-brk-id={item.brkId}
                  onClick={onSelect}
                  label={item.name}
                />
              </li>
            ))}
          </ul>
        </li>
      ))}
    </Ul>
  );
});

Suggest.defaultProps = {
  className: ``,
  items: null,
};

Suggest.propTypes = {
  className: PropTypes.string,
  items: PropTypes.shape({
    Adressen: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        brkId: PropTypes.string,
        vboId: PropTypes.string,
        ligId: PropTypes.string,
      }),
    ),
    'Kadastrale objecten': PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        brkId: PropTypes.string,
        vboId: PropTypes.string,
        ligId: PropTypes.string,
      }),
    ),
  }),
  onSelect: PropTypes.func.isRequired,
};

export default Suggest;
