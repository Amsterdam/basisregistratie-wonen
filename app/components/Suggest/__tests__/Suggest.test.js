import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-styled-components';
import { ThemeProvider } from '@datapunt/asc-ui';
import Suggest from '..';

const items = {
  Adressen: [
    {
      name: 'FakeStreet 999X',
      vboId: null,
      ligId: null,
      brkId: '0363010000740956',
    },
    {
      name: 'FakeStreet 999Y',
      vboId: null,
      ligId: '0363010000740956',
      brkId: null,
    },
    {
      name: 'FakeStreet 999Z',
      vboId: '0363010000740958',
      ligId: null,
      brkId: null,
    },
  ],
};

describe('Suggest', () => {
  afterEach(cleanup);

  it('matches the snapshot', () => {
    const { container } = render(
      <ThemeProvider>
        <Suggest onSelect={() => {}} items={items} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render null', () => {
    const { container, rerender } = render(
      <ThemeProvider>
        <Suggest onSelect={() => {}} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toBeNull();

    rerender(
      <ThemeProvider>
        <Suggest onSelect={() => {}} items={{}} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toBeNull();
  });

  it('should render category names', () => {
    const { queryByText } = render(
      <ThemeProvider>
        <Suggest onSelect={() => {}} items={{ Adressen: [], 'Kadastrale Objecten': [] }} />
      </ThemeProvider>,
    );

    expect(queryByText('Adressen')).toBeInstanceOf(HTMLElement);
    expect(queryByText('Kadastrale Objecten')).toBeInstanceOf(HTMLElement);
  });

  it('sets data attributes', () => {
    const numItems = Object.keys(items.Adressen).length;

    const { container } = render(
      <ThemeProvider>
        <Suggest onSelect={() => {}} items={items} />
      </ThemeProvider>,
    );

    const anchors = container.getElementsByTagName('a');

    expect(anchors).toHaveLength(numItems);
    expect(anchors[0].dataset.brkId).not.toBeUndefined();
    expect(anchors[0].dataset.vboId).not.toBeUndefined();
  });

  it('should call onSelect handler', () => {
    const onSelect = jest.fn();

    const { getByText } = render(
      <ThemeProvider>
        <Suggest onSelect={onSelect} items={items} />
      </ThemeProvider>,
    );

    fireEvent(
      getByText('FakeStreet 999Y'),
      new MouseEvent('click', {
        bubbles: true,
      }),
    );

    expect(onSelect).toHaveBeenCalled();
  });
});
