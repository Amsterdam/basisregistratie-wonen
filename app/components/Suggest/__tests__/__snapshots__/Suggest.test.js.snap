// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggest matches the snapshot 1`] = `
.c3 {
  padding: 0 !important;
}

.c3:focus {
  outline: 4px solid #ffc90e !important;
}

.c3:hover,
.c3:focus {
  color: #ec0000;
  fill: #ec0000;
}

.c5 {
  vertical-align: baseline;
  margin-right: 5px;
}

.c1 {
  padding-left: 4px;
}

.c2 {
  margin-left: 4px !important;
}

.c0 {
  background: white;
  border-top: none;
  border: #767676 solid 1px;
  width: 100%;
}

.c0 li:first-of-type {
  margin-top: 8px !important;
}

.c4 {
  color: #004699;
  display: inline-block;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  font-family: inherit;
  font-weight: 300;
  font-size: 1rem;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
  line-height: 1;
  font-style: normal;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c4:focus {
  outline: none;
  background-color: #FEC813;
}

.c4:hover {
  color: #ec0000;
}

.c6 {
  color: #004699;
  display: inline-block;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  font-family: inherit;
  font-weight: 300;
  font-size: 1rem;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
  line-height: 1;
  font-style: normal;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c6:focus {
  outline: none;
  background-color: #FEC813;
}

.c6:hover {
  color: #ec0000;
}

.c7 {
  color: #004699;
  display: inline-block;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  font-family: inherit;
  font-weight: 300;
  font-size: 1rem;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
  line-height: 1;
  font-style: normal;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c7:focus {
  outline: none;
  background-color: #FEC813;
}

.c7:hover {
  color: #ec0000;
}

<ul
  class=" links c0"
>
  <li>
    <strong
      class="c1"
    >
      Adressen
    </strong>
    <ul
      class="links"
    >
      <li>
        <a
          class="c2 c3 c4"
          data-brk-id="0363010000740956"
          data-vbo-id="0363010000740956"
          href="/"
        >
          <img
            class="c5"
            width="14"
          />
          <span
            class="linklabel"
          >
            FakeStreet 999X
          </span>
        </a>
      </li>
      <li>
        <a
          class="c2 c3 c6"
          data-brk-id="0363010000740957"
          data-vbo-id="0363010000740957"
          href="/"
        >
          <img
            class="c5"
            width="14"
          />
          <span
            class="linklabel"
          >
            FakeStreet 999Y
          </span>
        </a>
      </li>
      <li>
        <a
          class="c2 c3 c7"
          data-brk-id="0363010000740958"
          data-vbo-id="0363010000740958"
          href="/"
        >
          <img
            class="c5"
            width="14"
          />
          <span
            class="linklabel"
          >
            FakeStreet 999Z
          </span>
        </a>
      </li>
    </ul>
  </li>
</ul>
`;
